gpkit/model.py

```diff
@@ -1,250 +1,424 @@
-            If the substitutions entry is a single key, val holds the value
-        frombase : string (optional)
-            Which model state to update. By default updates the current state.
-        replace : boolean (optional, default is True)
-            Whether the substitution should only subtitute currently
-            unsubstituted values (False) or should also make replacements of
-            current substitutions (True).
-
-        Returns
-        -------
-        No return value: the model is modified in place.
-        """
-        # look for sweep variables
-        found_sweep = []
+        self.unsubbed_cs = np.hstack((mag(p.cs) for p in posynomials))
+        self.unsubbed_exps = functools_reduce(add, (p.exps for p in posynomials))
+        (self.unsubbed_varlocs,
+         self.unsubbed_varkeys) = locate_vars(self.unsubbed_exps)
+
+        subs = {var: var.descr["value"]
+                for var in self.variables if "value" in var.descr}
+        subs.update(self.substitutions)
+
+        (sweep, linkedsweep,
+         constants) = separate_subs(subs, self.unsubbed_varkeys,
+                                    self.unsubbed_varlocs)
+        solution = SolutionArray()
+        kwargs.update({"solver": solver})
+        kwargs.update({"verbosity": verbosity - 1})
+
+        if sweep:
+            if len(sweep) == 1:
+                sweep_grids = np.array(list(sweep.values()))
+            else:
+                sweep_grids = np.meshgrid(*list(sweep.values()))
+
+            N_passes = sweep_grids[0].size
+            sweep_vects = {var: grid.reshape(N_passes)
+                           for (var, grid) in zip(sweep, sweep_grids)}
+
+            if verbosity > 0:
+                print("Solving for %i variables over %i passes." % (
+                      len(self.variables), N_passes))
+
+            def solve_pass(i):
+                this_pass = {var: sweep_vect[i]
+                             for (var, sweep_vect) in sweep_vects.items()}
+                linked = {var: fn(*[this_pass[VarKey(v)]
+                                    for v in var.descr["args"]])
+                          for var, fn in linkedsweep.items()}
+                this_pass.update(linked)
+                constants_ = constants
+                constants_.update(this_pass)
+                program, subs, mmaps = self.formProgram(programType, posynomials,
+                                                        constants_)
+
+                try:
+                    result = program.solve(*args, **kwargs)
+                    sol = self.parse_result(result, subs, mmaps,
+                                            sweep, linkedsweep)
+                    return program, sol
+                except (RuntimeWarning, ValueError):
+                    return program, program.result
+
+            if POOL:
+                mapfn = POOL.map_sync
+            else:
+                mapfn = map
+
+            self.program = []
```

Comment:
so self.program is either a type `list` or a type `GeometricProgram`, depending on whether there are sweeps involved?